src/Codec/Xlsx/Parser/Internal/Util.hs

```diff
@@ -1,26 +1,43 @@
-{-# LANGUAGE OverloadedStrings #-}
 module Codec.Xlsx.Parser.Internal.Util
   ( boolean
+  , eitherBoolean
   , decimal
+  , eitherDecimal
   , rational
+  , eitherRational
   ) where
 
 import Data.Text (Text)
+import Control.Monad.Fail (MonadFail)
 import qualified Data.Text as T
 import qualified Data.Text.Read as T
+import qualified Control.Monad.Fail as F
 
-decimal :: (Monad m, Integral a) => Text -> m a
-decimal t = case T.signed T.decimal $ t of
+decimal :: (MonadFail m, Integral a) => Text -> m a
+decimal = fromEither . eitherDecimal
+
+eitherDecimal :: (Integral a) => Text -> Either String a
+eitherDecimal t = case T.signed T.decimal t of
   Right (d, leftover) | T.null leftover -> return d
-  _ -> fail $ "invalid decimal" ++ show t
+  _ -> Left $ "invalid decimal" ++ show t
+
+rational :: (MonadFail m) => Text -> m Double
+rational = fromEither . eitherRational
 
-rational :: Monad m => Text -> m Double
-rational t = case T.signed T.rational t of
+eitherRational :: Text -> Either String Double
+eitherRational t = case T.signed T.rational t of
   Right (r, leftover) | T.null leftover -> return r
```

Comment:
Just a nitpicking but probably it's better to be explicit and use `Right` instead of return if we switch to Either here.